// ts-openapi/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var joi_1 = __importDefault(require("joi"));
exports.Joi = joi_1.default;
var openapi_1 = require("./openapi/openapi");
exports.OpenApi = openapi_1.OpenApi;
var auth_1 = require("./openapi/helpers/auth");
exports.basicAuth = auth_1.basicAuth;
exports.apiKeyAuth = auth_1.apiKeyAuth;
exports.bearerAuth = auth_1.bearerAuth;
exports.cookieAuth = auth_1.cookieAuth;
exports.oauth2ImplicitAuth = auth_1.oauth2ImplicitAuth;
exports.oauth2AuthorizationCodeAuth = auth_1.oauth2AuthorizationCodeAuth;
exports.oauth2PasswordAuth = auth_1.oauth2PasswordAuth;
exports.oauth2ClientCredentialsAuth = auth_1.oauth2ClientCredentialsAuth;
var body_mimetype_1 = require("./openapi/helpers/body-mimetype");
exports.textPlain = body_mimetype_1.textPlain;
var types_1 = require("./openapi/helpers/types");
exports.Types = types_1.Types;
var openapi_mingle_1 = require("./openapi/openapi-mingle");
exports.OpenApiMingle = openapi_mingle_1.OpenApiMingle;
