// ts-openapi/lib/openapi/openapi.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ParameterIn;
(function (ParameterIn) {
    ParameterIn["Query"] = "query";
    ParameterIn["Path"] = "path";
    ParameterIn["Header"] = "header";
    ParameterIn["Cookie"] = "cookie";
    ParameterIn["Body"] = "requestBody";
})(ParameterIn = exports.ParameterIn || (exports.ParameterIn = {}));
var StringFormats;
(function (StringFormats) {
    StringFormats["DateTime"] = "date-time";
    StringFormats["Date"] = "date";
    StringFormats["Byte"] = "byte";
    StringFormats["Binary"] = "binary";
    StringFormats["Email"] = "email";
    StringFormats["Password"] = "password";
    StringFormats["Uuid"] = "uuid";
    StringFormats["Uri"] = "uri";
    StringFormats["Hostname"] = "hostname";
    StringFormats["Ipv4"] = "ipv4";
    StringFormats["Ipv6"] = "ipv6";
})(StringFormats = exports.StringFormats || (exports.StringFormats = {}));
var NumberFormats;
(function (NumberFormats) {
    NumberFormats["Float"] = "float";
    NumberFormats["Double"] = "double";
})(NumberFormats = exports.NumberFormats || (exports.NumberFormats = {}));
var IntegerFormats;
(function (IntegerFormats) {
    IntegerFormats["Int32"] = "int32";
    IntegerFormats["Int64"] = "int64";
})(IntegerFormats = exports.IntegerFormats || (exports.IntegerFormats = {}));
